File: src/GR-version/TI_chess.py
from ti_draw import set_color, fill_rect, draw_rect, clear, draw_text
from ti_image import load_image, show_image
from ti_system import wait_key #The screen is 318x212
from time import sleep

lastmovedelta : int = 0
lastmovetarget : int = 0
white_to_move : bool = True
piecesLayout : list[str|None] = []

select_index : int = 56
selected_index : int|None = None

legal_moves : list[list[int|str]] = []

# Fontend
def set_right_color(i) -> None:
    row=i//8+1
    if (i%2==0 if row%2==0 else i%2==1):
        set_color(121,92,52)
        return
    set_color(228,217,202)

def fill_square(i) -> None:
    pos = (59+(i%8)*25, 6+(i//8)*25)
    fill_rect(pos[0]-1, pos[1]-1, 27, 27)

def place_piece(name, i) -> None:
    load_image(name)
    show_image(63+25*(i%8), 10+25*(i//8))

def select_init() -> None:
    global select_index, selected_index
    if select_index == selected_index:
        set_color(0,0,255)
    else:
        set_right_color(select_index)
    draw_rect(59+(select_index%8)*25, 6+(select_index//8)*25, 25, 25)
    set_color(0,255,255)

def make_move(move) -> None:
    global lastmovetarget, lastmovedelta, white_to_move
    piecesLayout[move[1]] = piecesLayout[move[0]]
    piecesLayout[move[0]] = None
    fill_square(move[0])
    fill_square(move[1])
    place_piece('{}P'.format('W' if white_to_move else 'B'), move[1])
    if move[2] == 'e':
        ep_vanish = move[1] + 8*(1 if white_to_move else -1) # The pawn gets sent to another dimension
        piecesLayout[ep_vanish] = None
        fill_square(ep_vanish)
    lastmovedelta = abs(move[1]-move[0])
    lastmovetarget = move[1]
    if move[1]//8 == (0 if white_to_move else 7):
        print('{} won!'.format("White" if white_to_move else "Black"))
        sleep(3)
        init()
        return
    white_to_move = not white_to_move
    get_all_legal_moves(white_to_move)

def askCorrect() -> bool:
    set_color(255,255,255)
    draw_text(156,196,"Valid?")
    set_color(255,150,150)
    draw_text(16,196,"Yes")
    set_color(150,255,150)
    draw_text(256,196,"No")
    wait_key()
    if key == 68: # graph/f5 (key for 'yes')
        return True
    return False

def clearAsk():
    set_color(87,87,87)
    fill_rect(0,196,400,40)

# Backend
def get_all_legal_moves(white:bool):
    global legal_moves
    legal_moves = []
    multiplier = -1 + 2 * white
    opponent = not white
    i = 0
    for piece in piecesLayout:
        if piece is None or piece == opponent:
            i += 1
            continue
        i8m, i16m = i+8*multiplier, i+16*multiplier
        if i%8 > 0:
            if piecesLayout[i-1] == opponent and lastmovedelta == 16 and lastmovetarget == i-1:
                legal_moves.append([i,i8m-1,'e'])
            if piecesLayout[i8m-1] == opponent:
                legal_moves.append([i,i8m-1,None])
        if i%8 < 7:
            if piecesLayout[i+1] == opponent and lastmovedelta == 16 and lastmovetarget == i+1:
                legal_moves.append([i,i8m+1,'e'])
            if piecesLayout[i8m+1] == opponent:
                legal_moves.append([i,i8m+1,None])
        if piecesLayout[i8m] is None:
            legal_moves.append([i,i8m,None])
            if i//8 == (6 if white else 1):
                if piecesLayout[i16m] is None:
                    legal_moves.append([i,i16m,None])
        i += 1

def init():
    global piecesLayout, select_index, selected_index, white_to_move, lastmovetarget
    piecesLayout = [None, None, None, None, None, None, None, None, False, False, False, False, False, False, False, False, None, None, None, None, None, None, None, None,None, None, None, None, None, None, None, None,None, None, None, None, None, None, None, None,None, None, None, None, None, None, None, None, True, True, True, True, True, True, True, True, None, None, None, None, None, None, None, None]
    select_index = 56
    selected_index = None
    white_to_move = True
    lastmovetarget = 0
    get_all_legal_moves(white_to_move)
    clear()
    set_color(87,87,87)
    fill_rect(0,0,400,400)
    draw_rect(0,0,400,400)
    for i in range(64):
        fill_square(i)
    i = 0
    for img in piecesLayout:
        if img is not None:
            place_piece("WP" if img else "BP", i)
        i += 1
    set_color(0,255,255)
    draw_rect(59,181,25,25)

init()
    
while True:
    key = wait_key()
    if key == 1 and select_index%8 != 7: # Right arrow
        select_init()
        draw_rect(59+((select_index+1)%8)*25, 6+((select_index+1)//8)*25, 25, 25)
        select_index = select_index + 1
    elif key == 2 and select_index%8 != 0: # Left arrow
        select_init()
        draw_rect(59+((select_index-1)%8)*25, 6+((select_index-1)//8)*25, 25, 25)
        select_index = select_index - 1
    elif key == 3 and select_index//8 != 0: # Up arrow
        select_init()
        draw_rect(59+((select_index-8)%8)*25, 6+((select_index-8)//8)*25, 25, 25)
        select_index = select_index - 8
    elif key == 4 and select_index//8 != 7: # Down arrow
        select_init()
        draw_rect(59+((select_index+8)%8)*25, 6+((select_index+8)//8)*25, 25, 25)
        select_index = select_index + 8
    elif key == 5: # Enter
        # (un)select the current square
        if selected_index != None:
            set_right_color(selected_index)
            fill_rect(59+(selected_index%8)*25, 6+(selected_index//8)*25, 25, 25)
            possible_moves = [[selected_index, select_index, None], [selected_index, select_index, 'e']]
            if legal_moves.count(possible_moves[0]) > 0:
                correct = askCorrect()
                clearAsk()
                if not correct:
                    continue
                make_move(possible_moves[0])
            elif legal_moves.count(possible_moves[1]) > 0:
                correct = askCorrect()
                clearAsk()
                if not correct:
                    continue
                make_move(possible_moves[1])
            else:
                place_piece(piecesLayout[selected_index], selected_index)
                if select_index != selected_index:
                    draw_rect(59+(selected_index%8)*25, 6+(selected_index//8)*25, 25, 25)
            selected_index = None
        else:
            if piecesLayout[select_index] == white_to_move:
                selected_index = select_index
                set_color(0,0,255)
                fill_rect(59+(selected_index%8)*25, 6+(selected_index//8)*25, 25, 25)
                place_piece(piecesLayout[select_index], select_index)

File: src/GR-version/Cpp_Version/src/TI_chess.cpp
/**
 * @file TI_chess.cpp
 * @authors
 * Jens Boomgaard
 * Sebastiaan van Straten
 * Patrick Vreeburg
 * @brief The C++ version of TI-chess
 * @version 0.1
 * @date 2024-03-30
 * 
 * @copyright Copyright (c) 2024
 * 
 */
#include <stdint.h>
#include <graphx.h>
#include <fileioc.h>
#include <intce.h>
#include <ti/getcsc.h>
#include <ti/ui.h>
#include <sys/timers.h>

#include "gfx/vargfx.h"

#define NULL_VALUE -1

#define DEBUG false

const uint8_t BG_COLOR = 210;

const uint8_t SQUARE_WIDTH = 27;
const uint8_t BOARD_UPPER_LEFT_X = GFX_LCD_WIDTH / 2 - 4 * SQUARE_WIDTH;
const uint8_t BOARD_UPPER_LEFT_Y = GFX_LCD_HEIGHT / 2 - 4 * SQUARE_WIDTH;

const uint8_t PIECE_SIZE = 17;
const uint8_t PIECE_OFFSET = (SQUARE_WIDTH - PIECE_SIZE) / 2;

uint8_t lastMoveDelta = 0;
uint8_t lastMoveTarget = 0;

bool white_to_move = true;

uint64_t whiteBitboard = 0x0;
uint64_t blackBitboard = 0x0;
uint64_t totalBitboard = 0x0;

int16_t selectIndex = 0;
int16_t selectedIndex = 0;

// 30 legal moves is the max for only pawns
const uint8_t MAX_LEGAL_MOVES = 40;
int16_t legalMoves[MAX_LEGAL_MOVES][3];

enum moveType {
  NORMAL = 0,
  EN_PASSANT
};

// Get the pixel length of the strings in askCorrect()
const uint8_t VALID_LENGTH = gfx_GetCharWidth('V') + gfx_GetCharWidth('a') + gfx_GetCharWidth('l') + gfx_GetCharWidth('i') + gfx_GetCharWidth('d') + gfx_GetCharWidth('?');
const uint8_t YES_LENGTH = gfx_GetCharWidth('Y') + gfx_GetCharWidth('e') + gfx_GetCharWidth('s');
const uint8_t NO_LENGTH = gfx_GetCharWidth('N') + gfx_GetCharWidth('o');

const uint8_t WON_WIDTH = gfx_GetCharWidth('W') + gfx_GetCharWidth('h') + gfx_GetCharWidth('i') + gfx_GetCharWidth('t') + gfx_GetCharWidth('e') + gfx_GetCharWidth(' ') + gfx_GetCharWidth('w') + gfx_GetCharWidth('o') + gfx_GetCharWidth('n') + gfx_GetCharWidth('!');

// Function declarations
void setRightColor(const uint8_t i);
inline void fillSquare(const uint8_t i);
inline void placePiece(const bool white, const uint8_t i);
void selectInit();
void makeMove(const int16_t move[3]);
bool askCorrect();
void clearAsk();
void getAllLegalMoves(const bool white);
void init();

inline bool checkOverlap(const uint64_t& board, const uint8_t i);
int16_t movePossible(const uint8_t from, const uint8_t to);

int main() {

  if (!VARGFX_init()) {
    return 1;
  }
  
  int_Disable();

  gfx_Begin();

  os_RunIndicOff();

  // gfx_SetPalette(global_palette, sizeof_global_palette, 0);
  gfx_palette[208] = gfx_RGBTo1555(121, 92, 52);
  gfx_palette[209] = gfx_RGBTo1555(228, 217, 202);
  gfx_palette[210] = gfx_RGBTo1555(87, 87, 87);
  gfx_SetTransparentColor(5);

  init();

  while (true) {

    uint8_t key;

    while (!(key = os_GetCSC())) {};

    if (key == sk_Left && selectIndex % 8 != 7) {
      selectInit();
      gfx_Rectangle_NoClip(
        BOARD_UPPER_LEFT_X + SQUARE_WIDTH * ( 7 - (selectIndex + 1) % 8 ),
        BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * ( 7 - (selectIndex + 1) / 8 ),
        SQUARE_WIDTH,
        SQUARE_WIDTH
      );
      ++selectIndex;
    } else if (key == sk_Right && selectIndex % 8 != 0) {
      selectInit();
      gfx_Rectangle_NoClip(
        BOARD_UPPER_LEFT_X + SQUARE_WIDTH * ( 7 - (selectIndex - 1) % 8 ),
        BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * ( 7 - (selectIndex - 1) / 8 ),
        SQUARE_WIDTH,
        SQUARE_WIDTH
      );
      --selectIndex;
    } else if (key == sk_Down && selectIndex / 8 != 0) {
      selectInit();
      gfx_Rectangle_NoClip(
        BOARD_UPPER_LEFT_X + SQUARE_WIDTH * ( 7 - (selectIndex - 8) % 8 ),
        BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * ( 7 - (selectIndex - 8) / 8 ),
        SQUARE_WIDTH,
        SQUARE_WIDTH
      );
      selectIndex -= 8;
    } else if (key == sk_Up && selectIndex / 8 != 7) {
      selectInit();
      gfx_Rectangle_NoClip(
        BOARD_UPPER_LEFT_X + SQUARE_WIDTH * ( 7 - (selectIndex + 8) % 8 ),
        BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * ( 7 - (selectIndex + 8) / 8 ),
        SQUARE_WIDTH,
        SQUARE_WIDTH
      );
      selectIndex += 8;
    } else if (key == sk_Enter) {
      
      if (DEBUG) {
        gfx_FillRectangle_NoClip(0, 0, GFX_LCD_WIDTH, BOARD_UPPER_LEFT_Y);
        gfx_PrintStringXY(
          ((checkOverlap(whiteBitboard, selectIndex) && white_to_move) || (checkOverlap(blackBitboard, selectIndex) && !white_to_move)) ? "Yes" : "No"
        , 0, 0);
      }

      if (selectedIndex == NULL_VALUE) {

        if ( !((checkOverlap(whiteBitboard, selectIndex) && white_to_move) || (checkOverlap(blackBitboard, selectIndex) && !white_to_move)) ) {
          continue;
        }

        selectedIndex = selectIndex;
        gfx_SetColor(24);
        gfx_FillRectangle_NoClip(
          BOARD_UPPER_LEFT_X + SQUARE_WIDTH * (7 - selectedIndex % 8),
          BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * (7 - selectedIndex / 8),
          SQUARE_WIDTH,
          SQUARE_WIDTH
        );
        placePiece(checkOverlap(whiteBitboard, selectedIndex), selectedIndex);

      } else {

        setRightColor(selectedIndex);
        gfx_FillRectangle_NoClip(
          BOARD_UPPER_LEFT_X + SQUARE_WIDTH * (7 - selectedIndex % 8),
          BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * (7 - selectedIndex / 8),
          SQUARE_WIDTH,
          SQUARE_WIDTH
        );

        int16_t moveType = movePossible(selectedIndex, selectIndex);

        if (moveType == NULL_VALUE) {
          placePiece(checkOverlap(whiteBitboard, selectedIndex), selectedIndex);
          if (selectIndex != selectedIndex) {
            gfx_Rectangle_NoClip(
              BOARD_UPPER_LEFT_X + SQUARE_WIDTH * (7 - selectedIndex % 8),
              BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * (7 - selectedIndex / 8),
              SQUARE_WIDTH,
              SQUARE_WIDTH
            );
          }
        } else {
          bool correct = askCorrect();
          clearAsk();
          if (!correct) {
            setRightColor(selectedIndex);
            fillSquare(selectedIndex);
            placePiece(white_to_move, selectedIndex);
            selectedIndex = NULL_VALUE;
            continue;
          }
          int16_t move[3] = {selectedIndex, selectIndex, moveType};
          makeMove(move);
        }

        selectedIndex = NULL_VALUE;

      }
    } else if (key == sk_Clear) {
      break;
    }
  }

  os_RunIndicOn();

  gfx_End();

  int_Enable();

  return 0;
}

bool checkOverlap(const uint64_t& board, const uint8_t i) {
  return ( (board >> i) & 1 );
}

int16_t movePossible(const uint8_t from, const uint8_t to) {
  for (uint8_t i = 0; i < MAX_LEGAL_MOVES; ++i) {
    int16_t* move = legalMoves[i];
    if (move[0] == from && move[1] == to) {
      return move[2];
    }
  }
  return NULL_VALUE;
}

void setRightColor(const uint8_t i) {
  const uint8_t row = i/8 + 1;
  if ( (row % 2 == 0) ? (i % 2 == 0) : (i % 2 == 1) ) {
    gfx_SetColor(208); // (121, 92, 52) in RGB
    return;
  }
  gfx_SetColor(209); // (228, 217, 202) in RGB
}

void fillSquare(const uint8_t i) {
  gfx_FillRectangle_NoClip(
    BOARD_UPPER_LEFT_X + SQUARE_WIDTH * (7 - i % 8),
    BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * (7 - i / 8),
    SQUARE_WIDTH,
    SQUARE_WIDTH
  );
}

void placePiece(const bool white, const uint8_t i) {
  gfx_TransparentSprite_NoClip(
    (white) ? wp : bp,
    BOARD_UPPER_LEFT_X + PIECE_OFFSET + SQUARE_WIDTH * (7 - i % 8),
    BOARD_UPPER_LEFT_Y + PIECE_OFFSET + SQUARE_WIDTH * (7 - i / 8)
  );
}

void selectInit() {
  if (selectIndex == selectedIndex) {
    gfx_SetColor(24); // Blue
  } else {
    setRightColor(selectIndex);
  }
  gfx_Rectangle_NoClip(
    BOARD_UPPER_LEFT_X + SQUARE_WIDTH * (7 - selectIndex % 8),
    BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * (7 - selectIndex / 8),
    SQUARE_WIDTH,
    SQUARE_WIDTH
  );
  gfx_SetColor(31); // Cyan
}

// ! DEBUG
void printBitboard(const uint64_t& board) {
  gfx_FillRectangle_NoClip(0,0,GFX_LCD_WIDTH,BOARD_UPPER_LEFT_Y);
  gfx_SetTextFGColor(224);
  uint8_t x = 0;
  for (uint8_t i = 0; i < 64; ++i) {
    if (!checkOverlap(board, i)) continue;
    gfx_SetTextXY(x, 0);
    gfx_PrintUInt(i, 2);
    x += 20;
  }
  gfx_SetColor(0xff);
}

void makeMove(const int16_t move[3]) {
  uint64_t& player = (white_to_move) ? whiteBitboard : blackBitboard;
  uint64_t& opponent = (white_to_move) ? blackBitboard : whiteBitboard;

  if (DEBUG) {
    gfx_FillRectangle_NoClip(0,0,GFX_LCD_WIDTH,BOARD_UPPER_LEFT_Y);
    gfx_SetTextFGColor(31);
    gfx_SetTextXY(0,0);
    gfx_PrintInt(move[0], 2);
    gfx_SetTextXY(20,0);
    gfx_PrintInt(move[1], 2);
    gfx_SetTextXY(40,0);
    gfx_PrintInt(move[2], 1);
    sleep(2);
  }

  player += (1ull << move[1]);
  if ( checkOverlap(opponent, move[1]) ) opponent -= (1ull << move[1]);
  player -= (1ull << move[0]);
  
  if (DEBUG) printBitboard(player);

  setRightColor(move[0]);
  fillSquare(move[0]);
  setRightColor(move[1]);
  fillSquare(move[1]);

  placePiece(white_to_move, move[1]);

  if (move[2] == EN_PASSANT) { // En passant
    uint8_t ep_vanish = move[1] + 8 * ( (white_to_move) ? -1 : 1 );
    opponent -= (1ull << ep_vanish);
    setRightColor(ep_vanish);
    fillSquare(ep_vanish);
  }

  lastMoveDelta = (move[1] - move[0] > 0) ? (move[1] - move[0]) : (move[0] - move[1]);
  lastMoveTarget = move[1];

  if (move[1] / 8 == ((white_to_move) ? 7 : 0)) {
    gfx_SetTextFGColor(15);
    gfx_SetTextScale(2, 2);
    gfx_SetColor(0xff);
    gfx_FillRectangle_NoClip(0.5f * GFX_LCD_WIDTH - 1.1f * WON_WIDTH, 0.5f * GFX_LCD_HEIGHT - 0.1f * WON_WIDTH - 8, 2.2f * WON_WIDTH, 0.2f * WON_WIDTH + 16);
    gfx_PrintStringXY( (white_to_move) ? "White won!" : "Black won!", 0.5f * GFX_LCD_WIDTH - WON_WIDTH, 0.5f * GFX_LCD_HEIGHT - 8 );
    sleep(3);
    init();
    return;
  }

  white_to_move = !white_to_move;
  totalBitboard = whiteBitboard | blackBitboard;
  getAllLegalMoves(white_to_move);
  selectInit();
}

bool askCorrect() {
  gfx_SetTextFGColor(31); // White
  gfx_PrintStringXY("Valid?", 0.5 * GFX_LCD_WIDTH - 0.5 * VALID_LENGTH, GFX_LCD_HEIGHT - 10);
  gfx_SetTextFGColor(7); // Green
  gfx_PrintStringXY("Yes", 0.05 * GFX_LCD_WIDTH, GFX_LCD_HEIGHT - 10);
  gfx_SetTextFGColor(224); // Red
  gfx_PrintStringXY("No", 0.95 * GFX_LCD_WIDTH - NO_LENGTH, GFX_LCD_HEIGHT - 10);

  uint8_t key;

  while (!(key = os_GetCSC())) {};

  return (key == sk_Yequ);
}

void clearAsk() {
  gfx_SetColor(BG_COLOR);
  gfx_FillRectangle_NoClip(0, GFX_LCD_HEIGHT - 10, GFX_LCD_WIDTH, 10);
}

void setMove(int16_t from, int16_t to, moveType type, uint8_t& counter) {
  legalMoves[counter][0] = from;
  legalMoves[counter][1] = to;
  legalMoves[counter][2] = type;
  ++counter;
}

void getAllLegalMoves(const bool white) {

  for (uint8_t i = 0; i < MAX_LEGAL_MOVES; ++i) {
    legalMoves[i][0] = NULL_VALUE;
    legalMoves[i][1] = NULL_VALUE;
    legalMoves[i][2] = NULL_VALUE;
  }

  const int8_t MULTIPLIER = -1 + 2 * white;

  uint8_t counter = 0;

  const uint64_t& player = (white_to_move) ? whiteBitboard : blackBitboard;
  const uint64_t& opponent = (white_to_move) ? blackBitboard : whiteBitboard;

  for (uint8_t i = 0; i < 64; ++i) {
    if ( checkOverlap(opponent, i) || !checkOverlap(player, i) ) {
      continue;
    }

    const int16_t i8m = i + 8 * MULTIPLIER;
    const int16_t i16m = i + 16 * MULTIPLIER;

    if (i % 8 > 0) {
      if ( checkOverlap(opponent, i - 1) && lastMoveDelta == 16 && lastMoveTarget == i - 1 ) {
        setMove(i, i8m - 1, EN_PASSANT, counter);
      }
      if ( checkOverlap(opponent, i8m - 1) ) {
        setMove(i, i8m - 1, NORMAL, counter);
      }
    }
    if (i % 8 < 7) {
      if ( checkOverlap(opponent, i + 1) && lastMoveDelta == 16 && lastMoveTarget == i + 1 ) {
        setMove(i, i8m + 1, EN_PASSANT, counter);
      }
      if ( checkOverlap(opponent, i8m + 1) ) {
        setMove(i, i8m + 1, NORMAL, counter);
      }
    }
    if (checkOverlap(totalBitboard, i8m)) {
      continue;
    }
    setMove(i, i8m, NORMAL, counter);
    if ( checkOverlap(totalBitboard, i16m) || i / 8 == ( (white_to_move) ? 6 : 1 ) ) {
      continue;
    }
    setMove(i, i16m, NORMAL, counter);
  }
}

void init() {
  whiteBitboard = 0xff00;
  blackBitboard = 0xff000000000000;
  totalBitboard = whiteBitboard | blackBitboard;
  selectIndex = 7;
  selectedIndex = NULL_VALUE;
  white_to_move = true;
  lastMoveTarget = 0;
  getAllLegalMoves(true);
  
  // Draw the board
  gfx_FillScreen(BG_COLOR);
  for (uint8_t i = 0; i < 64; ++i) {
    setRightColor(i);
    fillSquare(i);

    if (checkOverlap(whiteBitboard, i)) {
      placePiece(true, i);
    } else if (checkOverlap(blackBitboard, i)) {
      placePiece(false, i);
    }
  }
  gfx_SetColor(31); // Cyan
  gfx_Rectangle_NoClip(
    BOARD_UPPER_LEFT_X + SQUARE_WIDTH * (7 - selectIndex % 8),
    BOARD_UPPER_LEFT_Y + SQUARE_WIDTH * (7 - selectIndex / 8),
    SQUARE_WIDTH,
    SQUARE_WIDTH
  );
}